<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">

    <title>MHcut Browser</title>

    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,700" rel="stylesheet">
    <link href="/style.css" rel="stylesheet">

    <script src="/dist/bundle.js"></script>
</head>
<body>
<div id="column-help-text"></div>
<div id="advanced-search-container" class="modal-container">
    <div id="advanced-search-modal" class="modal">
        <button class="close-modal"><span class="material-icons">close</span></button>
        <div id="advanced-search-modal-scroll-container" class="modal-scroll-container">
            <h2>
                Advanced Search
                <button id="toggle-advanced-search-help" class="toggle-modal-help">
                    <span class="material-icons">help</span>
                </button>
            </h2>
            <div id="advanced-search-help">
                <p>
                    The advanced search mechanism allows for complex filtering of variants in the database by allowing
                    the user to create a <strong>set of rules</strong> that determine if a variant entry is or is not
                    allowed the result set.
                </p>
                <p>
                    To begin, add at least one condition with the "Add Condition" button in the bottom left of the modal
                    window. This adds a set of fields which can be used to create a <strong>single rule</strong>
                    applying to a <strong>specific column</strong> in the database. The options for the new rule are
                    broken down as follows: 1) A joining <em>boolean</em> operator (not present on the first operator).
                    2) An optional negation field that takes the rest of the condition and inverts it. For example, if
                    the current condition is "gene_info contains 'LOC'", then this would turn the condition into
                    "gene_info <strong>does not</strong> contain 'LOC'". 3) A field drop-down which specifies which
                    field the condition should apply to. 4) An operator for the condition; the available operators
                    depend on the type of data stored in the column. This relates the selected column with the value
                    entered into the text box on the right (5).
                </p>
                <p>
                    For example, take the <strong>gene_info</strong> field. This field allows 9 different operators:
                    equals, 4 inequality operators, starts/ends with, and contains. In this case, the
                    <strong>contains</strong> operator is helpful. If the string "LOC" is being searched in the
                    <strong>gene_info</strong> field, the <strong>contains</strong> operator would yield positive
                    matches with, for example, "<em>LOC</em>105378947:105378947" or
                    "KAZN:23254|<em>LOC</em>107985467:107985467". If the <strong>starts_with</strong> operator was used,
                    only the first example would match.
                </p>
                <p>
                    The <strong>equals</strong> operator is available for all data types and only yields direct matches,
                    meaning that the contents of the column entries must match exactly (with no room for any characters,
                    including whitespace, on either side of a value). The inequality operators use the same order as
                    when sorting the columns, so <strong>2</strong> < <strong>3</strong>, <strong>0.2</strong> <
                    <strong>0.3</strong>, and <strong>abc</strong> < <strong>abe</strong> < <strong>bce</strong>.
                </p>
                <p>
                    The joining boolean operators available are <strong>AND</strong> and <strong>OR</strong>. They
                    behave as follows, using "True" and "False" as results of the actual comparison operators described
                    above (so, for example, 1 = 1 is True and "hello" contains "a" is False). Note that the OR used in
                    this case is <strong>inclusive</strong> as opposed to exclusive.
                </p>
                <p>
                    True AND True is True; True AND False is False; <strong>True OR True is True</strong>;
                    True OR False is True; False AND/OR False is False.
                </p>
                <p>
                    In this way, multiple rules can be strung together into a query. When the query is built, an
                    encoded version can be saved by pressing the "Save" button, which will clear and re-populate the
                    search field with a machine-readable version of the query. The query can then be optionally combined
                    with "quick filters" in the sidebar before being used to filter the variants.
                </p>
            </div>
            <!--
            <p style="margin-top: 0;">
                <strong>Order of Boolean Operations:</strong>
                In chains of <strong>AND</strong> and <strong>OR</strong>, <strong>AND</strong> takes precedence. For
                example, the following are equivalent:
            </p>
            <div class="code-comparison">
                <code>
                    cond_1 <strong>OR</strong>
                    cond_2 <strong>AND</strong> cond_3 <strong>OR</strong>
                    cond_4 <strong>AND</strong> cond_5 <strong>AND</strong> cond_6</code>
                <code>
                    cond_1 <strong>OR</strong>
                    <strong>(</strong>cond_2 <strong>AND</strong> cond_3<strong>)</strong> <strong>OR</strong>
                    <strong>(</strong>cond_4 <strong>AND</strong> cond_5 <strong>AND</strong> cond_6<strong>)</strong>
                </code>
            </div>
            -->
            <p style="margin-top: 0;">
                For example to select exonic rare variants whose micro-homology starts with at least 3 exact matches,
                with at most 10 bp of non-homologous sequence and with at least one PAM with unique protospacer:
            </p>
            <div class="code-comparison">
                <code>
                    location equals exon <strong>AND</strong> topmed < 0.01 <strong>AND</strong> mh_1l >= 3
                    <strong>AND</strong> mh_dist <= 10 <strong>AND</strong> pam_uniq > 0
                </code>
            </div>

            <ul id="advanced-search-conditions"></ul>
        </div>
        <div id="advanced-search-modal-options" class="modal-options">
            <div>
                <button id="add-search-condition" class="with-icon">
                    <span class="material-icons">add</span> Add Condition
                </button>
            </div>
            <div>
                <button id="save-search-query" class="with-icon">
                    <span class="material-icons">save</span> Save
                </button>
            </div>
        </div>
    </div>
</div>
<div id="export-options-container" class="modal-container">
    <div id="export-options-modal" class="modal">
        <button class="close-modal"><span class="material-icons">close</span></button>
        <div class="modal-scroll-container">
            <h2>Export Options</h2>
            <p>
                <strong>Total Matching Variants:</strong> <span id="matching-variants-export">many</span> <br>
                <strong>Total Matching Guides:</strong> <span id="matching-guides-export">many</span>
            </p>
            <form>
                <label for="export-format">
                    Export Format:
                    <select name="export-format" id="export-format">
                        <option value="tsv">TSV</option>
                    </select>
                </label>
                <label for="guides-with-variant-info" id="label-guides-with-variants-info">
                    <input type="checkbox" id="guides-with-variant-info" name="guides-with-variant-info" checked>
                    Include variant information in guide rows? <em>(recommended)</em>
                </label>
            </form>
        </div>
        <div class="modal-options">
            <div style="flex: 0;"></div>
            <div>
                <button id="export-variants">Export Variants</button>
                <button id="export-guides">Export Guides</button>
                <button id="export-combined">Export Combined</button>
            </div>
        </div>
    </div>
</div>
<div id="variant-guides-container" class="modal-container">
    <div id="variant-guides-modal" class="modal">
        <button class="close-modal"><span class="material-icons">close</span></button>
        <div class="modal-scroll-container">
            <h2>Guides for Variant <span id="variant-for-guides">0</span></h2>
            <table id="variant-guides-table">
                <thead></thead>
                <tbody></tbody>
            </table>
        </div>
        <div class="modal-options">
            <div></div>
            <div><button id="export-variant-guides">Export</button></div>
        </div>
    </div>
</div>
<div id="terms-of-use-container" class="modal-container">
    <div id="terms-of-use-modal" class="modal">
        <button class="close-modal"><span class="material-icons">close</span></button>
        <div class="modal-scroll-container">
            <h2>MHcut Browser Terms of Use</h2>
            <div class="terms-of-use">
                <p>
                    The MHcut Code and Software are being provided under a Limited Copyright License for Research Use by
                    Non-Profit and Government Institutions (GNU GPLv3). The MHcut Code and Software are being provided
                    as a service for research, educational, instructional and non-commercial purposes only.
                </p>
                <p>
                    &ldquo;Code&rdquo; means the MHcut source code made available in the GitHub repository at
                    <a href="https://github.com/WoltjenLab/MHcut" target="_blank"
                       rel="noopener">https://github.com/WoltjenLab/MHcut</a>; and
                    &ldquo;Software&rdquo; means the code made available as a web service by the Canadian Centre for
                    Computational Genomics and the Center for iPS Cell Research and Application, Kyoto University, at
                    <a href="https://mhcut-browser.genap.ca/">https://mhcut-browser.genap.ca/</a>.
                </p>
                <p>
                    By using the Software or downloading the Code accompanying this license, you are consenting to be
                    bound by all the terms of this license listed below.
                </p>
                <ul>
                    <li>
                        You are an actively enrolled student, post-doctoral researcher, or faculty member at a
                        degree-granting educational institution or accredited government or not-for-profit research
                        institution; and You are hereby granted a non-exclusive, non-transferable license to use the
                        Code and/or Software solely for educational, instructional, and/or non-commercial
                        research purposes;
                    </li>
                    <li>
                        You understand that all results produced using the Code or Software may only be used for
                        non-commercial research and/or academic purposes;
                    </li>
                    <li>
                        You understand that to obtain any right to use the Code or Software for commercial purposes, or
                        in the context of industrially sponsored research, You must enter into an appropriate, separate
                        and direct license agreement with the Owners.
                    </li>
                    <li>You will not redistribute unmodified versions of the Code or Software;</li>
                    <li>
                        You will redistribute modifications, if any, under the same terms as this license and only to
                        educational, non-profits and government institutions;
                    </li>
                    <li>
                        You must credit the authors of the Code: Janin Grajcarek, Jean Monlong and Knut Woltjen and the
                        authors of the Software: David Lougheed and cite &ldquo;Genome-wide microhomologies enable
                        precise template-free editing of biologically relevant deletion mutations&rdquo;, Nature
                        Communications, 2019, doi:<…(fill in after publication)...>;
                    </li>
                    <li>
                        You acknowledge that the Software or Code have not been developed to meet your individual
                        requirements and that it is therefore your responsibility to ensure that the facilities and
                        functions of the Software or Code meet your requirements; and
                    </li>
                    <li>
                        You understand that neither the names of the Owners nor the names of the authors may be used to
                        endorse or promote products derived from the Code or Software without specific prior written
                        permission.
                    </li>
                </ul>
                <p>
                    &ldquo;Copyright 2019. Canadian Centre for Computational Genomics and the Center for iPS Cell
                    Research and Application, Kyoto University. All Rights Reserved.&rdquo;
                </p>
                <h3>Contact</h3>
                <p>
                    This web tool is hosted by the Canadian Center for Computational Genomics and was developed by the
                    <a href="http://www.cira.kyoto-u.ac.jp/woltjen/" target="_blank" rel="noopener">Woltjen lab</a> at
                    the Center for iPS Cell Research and Application (CiRA), Kyoto University, Kyoto, Japan in
                    collaboration with the
                    <a href="http://www.computationalgenomics.ca/BourqueLab/" target="_blank"
                       rel="noopener">Bourque lab</a> at the Department of Human Genetics, McGill University, Montreal,
                    Canada.
                </p>
                <p>
                    For more information and to report bugs, please contact
                    <a href="mailto:woltjen@cira.kyoto-u.ac.jp">Knut Woltjen</a>.
                </p>
            </div>
        </div>
    </div>
</div>
<div id="report-bug-container" class="modal-container">
    <div id="report-bug-modal" class="modal">
        <button class="close-modal"><span class="material-icons">close</span></button>
        <div class="modal-scroll-container">
            <h2>Report a Bug in the MHcut Browser</h2>
            <label for="report-bug-email">Your Email</label>
            <input type="email" id="report-bug-email" name="report-bug-email" required
                   placeholder="name@example.com">
            <label for="report-bug-text">Please Describe What Happened</label>
            <textarea name="report-bug-text" id="report-bug-text" rows="5" required></textarea>
        </div>
         <div class="modal-options">
            <div></div>
            <div><button id="report-bug-submit" type="submit">Submit Report</button></div>
        </div>
    </div>
</div>
<header id="main-header">
    <div id="developed-by">
        <span>Developed by</span>
        <a href="http://www.cira.kyoto-u.ac.jp/e/" target="_blank" rel="noopener noreferrer">
            <img src="/cira_logo.png" alt="CiRA">
        </a>
        <a href="http://www.computationalgenomics.ca/" target="_blank" rel="noopener noreferrer">
            <img src="/c3g_logo_small.png" alt="C3G">
        </a>
    </div>
    <div>
        <h1>MHcut Browser</h1>
    </div>
    <div>
        <button id="terms-of-use"><span class="material-icons">list_alt</span> Terms of Use</button>
        <button id="report-bug"><span class="material-icons">bug_report</span> Report a Bug</button>
    </div>
</header>
<main>
<nav id="table-view-nav">
    <div>
        <ul>
            <li class="active" id="view-variants">Variants</li>
            <li id="view-guides">Corresponding Guides</li>
        </ul>
    </div>
    <div>
        <button id="toggle-all-additional-columns">Show All Additional Columns</button>
    </div>
    <!--
    <div>
        <span class="help-text">Hover over column names to see description.</span>
    </div>
    -->
    <div style="flex: 1; text-align: right;">
        <button id="show-export">Export Matches...</button>
        <button id="sidebar-toggle" class="active" title="Sidebar Toggle">
            <span class="material-icons">vertical_split</span>
        </button>
    </div>
</nav>
<section id="table-display" class="loading">
    <table id="entry-table"><thead><tr class="group-row"></tr><tr class="header-row"></tr></thead><tbody></tbody></table>
</section>
<section id="table-pagination">
    <div>
        <label for="items-per-page">
            Items Per Page:
            <select name="items-per-page" id="items-per-page">
                <option value="25">25</option>
                <option value="50">50</option>
                <option value="100" selected>100</option>
                <option value="250">250</option>
                <option value="500">500</option>
                <option value="1000">1000</option>
                <option value="999999999999">All Matches</option>
            </select>
        </label>
    </div>
    <div>
        Page <span id="current-page">1</span> of <span id="total-pages">multiple</span>
        <span id="total-entries-text">
            (<span id="total-variants">many</span> variants, <span id="total-guides">many</span> guides)
        </span>
    </div>
    <div style="font-size: 0;">
        <button id="first-page" class="icon-button" disabled>
            <span class="material-icons">first_page</span>
        </button>
        <div class="button-row">
            <button id="prev-page" class="icon-button" disabled>
                <span class="material-icons">chevron_left</span>
            </button>
            <button id="next-page" class="icon-button" disabled>
                <span class="material-icons">chevron_right</span>
            </button>
        </div>
        <button id="last-page" class="icon-button" disabled>
            <span class="material-icons">last_page</span>
        </button>
    </div>
</section>
</main>
<aside id="filter-options">
    <!--<div id="app-logo-container">
        <img src="/logo.png" alt="MHcut Browser Logo" id="app-logo">
    </div>
    <h1>MHcut Browser</h1>
    <div id="developed-by">
        <span>developed by</span>
        <a href="http://www.cira.kyoto-u.ac.jp/e/" target="_blank" rel="noopener noreferrer">
            <img src="/cira_logo.png" alt="CiRA">
        </a>
        <a href="http://www.computationalgenomics.ca/" target="_blank" rel="noopener noreferrer">
            <img src="/c3g_logo_small.png" alt="C3G">
        </a>
    </div>
    <hr>-->
    <form id="filter-search-form">
        <h2><label for="search-query">SEARCH</label></h2>
        <input type="text" name="search-query" id="search-query" placeholder="Query all fields...">
        <span class="input-help-text">e.g. gene name, RS ID</span>
        <div class="row">
            <h3 style="margin: 0; padding-right: 0.6em; line-height: 39px;">OR</h3>
            <button id="show-advanced-search" type="button" class="col"
                    style="margin: 5px 0;">Advanced Search</button>
        </div>
        <hr>
        <h2>POSITION SEARCH</h2> <!-- TODO: LABEL-IFY -->
        <input type="text" name="position-query" id="position-query" placeholder="chr#:start-end">
        <span class="input-help-text">e.g. chr1:10109-10245</span>
        <hr>
        <h2>QUICK FILTERS</h2>
        <h3>Variant Location</h3>
        <div id="gene-location-checkboxes"></div>
        <hr>
        <label for="min-mh-1l">Minimum consecutive first match in MH</label>
        <input type="number" min="0" value="3" id="min-mh-1l" name="min-mh-1l">
        <hr>
        <label for="clinvar" class="checkbox-label">
            <input type="checkbox" id="clinvar" name="clinvar" checked> Variant in ClinVar
        </label>
        <hr>
        <label for="ngg-pam-available" class="checkbox-label">
            <input type="checkbox" id="ngg-pam-available" name="ngg-pam-available"> NGG PAM Available
        </label>
        <label for="unique-guide-available" class="checkbox-label">
            <input type="checkbox" id="unique-guide-available" name="unique-guide-available"> Unique Guide Avail.
        </label>
        <hr>
        <button id="apply-filters" type="submit" disabled>Apply Filters</button>
        <button id="clear-filters" type="button" disabled>Clear Filters</button>
        <hr>
        <p>
            <strong>Note:</strong> After applying filters, it may take time for the filter
            controls to become available again, due to calculating the number of matches.
        </p>
    </form>
</aside>
</body>
</html>
